### ISSUE TITLE
[PushNotificationIOS] popInitialNotification always returns null

### ISSUE THREAD

**Issue by tachim** [facebook/react-native#2658]
I followed the directions [here](https://medium.com/@DannyvanderJagt/how-to-use-push-notifications-in-react-native-41e8b14aadae) to set up push notifs. In my AppStateIOS 'active' change callback popInitialNotification always returns null even when the app is launched from a click on a push notif. 

For anyone else seeing this problem, a workaround is to use the PushNotificationIOS notification callback to add to a notif queue when the AppStateIOS.currentState is 'background' and then read from that queue in the 'active' change callback.

**Comment by brentvatne** [comment]
cc @mattmo

**Comment by PhilippKrone** [comment]
Any update here? I seem to have the same issue.

**Comment by PhilippKrone** [comment]
@tachim Can you share some code about how you have done your workaround in more detail?

**Comment by PhilippKrone** [comment]
@vjeux I see that you've written the original code concerning "popInitialNotification" using git blame. Are we making a mistake here or is there something wrong in the OSS version of RN here?

**Comment by vjeux** [comment]
I did not, I just did the export before we had proper tooling. cc @ericvicenti

**Comment by ericvicenti** [comment]
The intention is that `popInitialNotification` will only return the notif when the app cold start comes from a notification.

To get notifs once the app is running (even in the background), you must subscribe to them:

```
PushNotificationIOS.addEventListener(‘notification’, (notif) => { ... });
```

The docs aren't very clear on this, and improvements would be welcome!

Hopefully I understood you correctly. If that doesn't work then we probably have a bug somewhere.

**Comment by PhilippKrone** [comment]
@ericvicenti OK, i get your point. In my case, I'm subscribing to 'notification' and that's working pretty well as long as the app is in 'active' state. The moment the app goes into background (respectively a few seconds later), my code in the (notif) => { ... } function is no longer executed. That's why I thought that using "popInitialNotification" is the right way to get the payload of the PushNotification even if the app is opened from background (and not only as a cold start).

@tachim Do you experience the same behaviour?

**Comment by PhilippKrone** [comment]
To give more details, this function that I have in appdelegate.m is only triggered if the app is active (and not in the background):

`````javascript
- (void)application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)notification
{
    [RCTPushNotificationManager application:application didReceiveRemoteNotification:notification];
}
`````

**Comment by tachim** [comment]
@ericvicenti I see, so just to be clear by "cold start" you're not including "waking from background?"

**Comment by ericvicenti** [comment]
Right, the cold start is when the app is not already running in the background or the user has manually terminated it. 

There are some similarities with this SO question:

http://stackoverflow.com/questions/28526709/waking-up-app-from-background-ios

Notifs from `application:didReceiveRemoteNotification` gets passed to JS as a notification event. Notifs from `application:didFinishLaunchingWithOptions` are available in `popInitialNotification`

**Comment by PhilippKrone** [comment]
@ericvicenti I think then there might be bug somewhere. My current setup is:

1) I have the following code in my AppDelegate.m
````javascript

- (void)application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)notification
{
    [RCTPushNotificationManager application:application didReceiveRemoteNotification:notification];
}
`````
2) I can successfully receive PushNotifications as long as my app is active and can verify that by adding  a breakpoint in aforementioned code
3) This very breakpoint is not triggered if my app is in background (not terminated)
4) My app background modes are "fetch" and "remote-notification"

Do you have any idea here?

Regards
Philipp

**Comment by ide** [comment]
Apple is possibly filtering out your push notifications. I believe RN is unrelated in this case.

**Comment by PhilippKrone** [comment]
Do you know any possible reasons for that? The moment I bring my app in backhround, its not triggered anymore. If its back in foreground its working again.

**Comment by ide** [comment]
I think it has to do with your APNS payload and Apple's power conservation.

**Comment by PhilippKrone** [comment]
@ide: indeed - thanks a lot for your help!

**Comment by nautical** [comment]
I am facing same issue , on cold start i receive the value as null , when app opens from notification . When app is in background or is active it works fine .

**Comment by nautical** [comment]
Ok so I was facing the issue in both notification and linkingios ... the initial url / notification I was not able to access . For others who come here , what I ended up doing is adding the following code in didFinishLaunchingWithOptions 

    NSURL *url = (NSURL *)[launchOptions valueForKey:UIApplicationLaunchOptionsURLKey];

    NSDictionary *remoteNotif = [launchOptions objectForKey:
                                 UIApplicationLaunchOptionsRemoteNotificationKey];

    NSMutableDictionary *Mut_launchprops = [NSMutableDictionary dictionary];

    if(url){
        [Mut_launchprops setObject:[url absoluteString] forKey:@"url"];
    }
];
`
and adding mut_launchprops in initialProperties of RCTRootView .

**Comment by jsierles** [comment]
@facebook-github-bot answered

**Comment by joshuapinter** [comment]
@nautical Any idea if this only works for remote notifications? I'm getting `null` for all my cold start launches from local notifications.

Thanks!

**Comment by joshuapinter** [comment]
FYI, this is still an issue with React Native. I opened up another issue to capture this: #8580 

Inspired by @nautical's solution, which is a bit dated and didn't work for me with the React Native 0.28, I managed to get a solution that works brilliantly for doing a cold launch of a React Native app from a local push notification.

The solution is in Issue #8580 as well as this StackOverflow Answer: http://stackoverflow.com/a/38191351/293280.
